packager/VisibilityChecker.cc

```diff
@@ -0,0 +1,558 @@
+    void exportPackageRoots(core::GlobalState &gs) {
+        this->exportRoot(gs, core::Symbols::root());
+
+        auto test = core::Symbols::root().data(gs)->findMember(gs, core::Names::Constants::Test());
+        if (test.exists() && test.isClassOrModule()) {
+            this->exportRoot(gs, test.asClassOrModuleRef());
+        }
+    }
+
+    void setPackageLocs(core::MutableContext ctx, core::LocOffsets loc, core::ClassOrModuleRef sym) {
+        std::vector<core::NameRef> names;
+
+        while (sym.exists() && sym != core::Symbols::PackageSpecRegistry()) {
+            // The symbol isn't a package name if it's defined outside of the package registry.
+            if (sym == core::Symbols::root()) {
+                return;
+            }
+
+            names.emplace_back(sym.data(ctx)->name);
+            sym = sym.data(ctx)->owner;
+        }
+
+        absl::c_reverse(names);
+
+        auto &db = ctx.state.packageDB();
+
+        {
+            auto packageSym = core::Symbols::root();
+            for (auto name : names) {
+                auto member = packageSym.data(ctx)->findMember(ctx, name);
+                if (!member.exists() || !member.isClassOrModule()) {
+                    packageSym = core::Symbols::noClassOrModule();
+                    break;
+                }
+
+                packageSym = member.asClassOrModuleRef();
+            }
+
+            if (packageSym.exists()) {
+                auto file = packageSym.data(ctx)->loc().file();
+                if (db.getPackageNameForFile(file) != this->package.mangledName()) {
+                    packageSym.data(ctx)->addLoc(ctx, ctx.locAt(loc));
+                }
+            }
+        }
+
+        {
+            auto testSym = core::Symbols::root();
+            auto member = testSym.data(ctx)->findMember(ctx, core::Names::Constants::Test());
+            if (!member.exists() || !member.isClassOrModule()) {
+                return;
+            }
+
+            testSym = member.asClassOrModuleRef();
+            for (auto name : names) {
+                auto member = testSym.data(ctx)->findMember(ctx, name);
+                if (!member.exists() || !member.isClassOrModule()) {
+                    testSym = core::Symbols::noClassOrModule();
+                    break;
+                }
+
+                testSym = member.asClassOrModuleRef();
+            }
+
+            if (testSym.exists()) {
+                auto file = testSym.data(ctx)->loc().file();
+                if (db.getPackageNameForFile(file) != this->package.mangledName()) {
+                    testSym.data(ctx)->addLoc(ctx, ctx.locAt(loc));
+                }
+            }
+        }
+    }
+
+    void checkExportPackage(core::MutableContext &ctx, core::LocOffsets loc, core::NameRef litPackage) {
+        if (litPackage != this->package.mangledName() && !this->package.importsPackage(litPackage)) {
```

Comment:
I'm having a tough time understanding what the `&& !this->package.importsPackage(litPackage)` is for.

Wouldn't we want to always report the error if `litPackage != this->package.mangledName()`?

(It seems like the current behavior would allow something like

```ruby
class A::B < PackageSpec
  import A::X

  export A::X::PrivateFoo
end
```

but correct me if i'm wrong.

Reply:
Unfortunately this is to allow re-exporting constants from other packages, which the current system allows. I think it would be worthwhile removing this behavior, and disallowing re-exports.